src/udev/iocost/iocost.c

```diff
@@ -0,0 +1,244 @@
+                return log_oom();
+
+        model_key = strjoin("IOCOST_MODEL_", upper_name);
+        if (!model_key)
+                return log_oom();
+
+        r = sd_device_get_property_value(device, model_key, model);
+        if (r < 0)
+                return log_error_errno(r, "Model key missing from hwdb for iocost solution '%s': %m", name);
+
+        r = sd_device_get_property_value(device, qos_key, qos);
+        if (r < 0)
+                return log_error_errno(r, "QoS key missing from hwdb for iocost solution '%s': %m", name);
+
+        return 0;
+}
+
+static int apply_solution_for_path(const char *path, const char *name) {
+        _cleanup_(sd_device_unrefp) sd_device *device = NULL;
+        _cleanup_free_ char *model = NULL, *qos = NULL;
+        const char *model_params = NULL, *qos_params = NULL;
+        dev_t devnum;
+        int r;
+
+        r = sd_device_new_from_path(&device, path);
+        if (r < 0)
+                return log_error_errno(r, "Error looking up device: %m");
+
+        r = query_named_solution(device, name, &model_params, &qos_params);
+        if (r < 0)
+                return r;
+
+        r = sd_device_get_devnum(device, &devnum);
+        if (r < 0)
+                return log_error_errno(r, "Error getting devnum for device %s: %m", path);
+
+        if (asprintf(&model, DEVNUM_FORMAT_STR " model=linear ctrl=user %s", DEVNUM_FORMAT_VAL(devnum), model_params) < 0)
+                return log_oom();
+
+        if (asprintf(&qos, DEVNUM_FORMAT_STR " enable=1 ctrl=user %s", DEVNUM_FORMAT_VAL(devnum), qos_params) < 0)
+                return log_oom();
+
+        log_debug("Applying iocost parameters to %s using solution '%s'\n\tio.cost.model: %s\n\tio.cost.qos: %s\n", path, name ? name : "default", model, qos);
+
+        r = cg_set_attribute("io", NULL, "io.cost.qos", qos);
+        if (r < 0) {
+                log_full_errno(r == -ENOENT ? LOG_DEBUG : LOG_ERR, r, "Failed to set io.cost.qos: %m");
+                return r == -ENOENT ? 0 : r;
+        }
+
+        r = cg_set_attribute("io", NULL, "io.cost.model", model);
+        if (r < 0) {
+                log_full_errno(r == -ENOENT ? LOG_DEBUG : LOG_ERR, r, "Failed to set io.cost.model: %m");
+                return r == -ENOENT ? 0 : r;
+        }
+
+        return 0;
+}
+
+static int query_solutions_for_path(const char *path) {
+        _cleanup_(sd_device_unrefp) sd_device *device = NULL;
+        _cleanup_strv_free_ char **solutions = NULL;
+        const char *model_name = NULL;
+        int r;
+
+        r = sd_device_new_from_path(&device, path);
+        if (r < 0)
+                return log_error_errno(r, "Error looking up device: %m");
+
+        if ((sd_device_get_property_value(device, "ID_MODEL_FROM_DATABASE", &model_name) < 0 &&
+             sd_device_get_property_value(device, "ID_MODEL", &model_name) < 0)) {
+                log_info("Model name for device %s is unknown", path);
+                model_name = "unknown";
```

Comment:
shouldn't we fail here?

Reply:
I believe my thinking when I wrote it this way was that one could add a hwdb file for a device path even with a missing model, but I think this should never happen, so fail it is.